### PR TITLE
Clarify code deploy behavior with regards to downstream tasks

### PR DESCRIPTION
**Pull request by jwitz**
Resolves https://github.com/astronomer/docs/issues/323

Added a recommendation based on the discussion in the original issue (adding retries at the DAG level) lmk if there's anything else I should include here.

### CODE REVIEW

File: cloud/deploy-code.md [hunk 1]
```diff
@@ -56,16 +56,18 @@ Once you log in, you should see the DAGs you just deployed.
...
 
 ![Deploy Code](/img/docs/deploy-architecture.png)
 
 If you deploy code to a Deployment that is running a previous version of your code, then the following happens:
 
-1. Tasks that are `running` will continue to execute on existing Celery Workers and will not be interrupted unless the task does not complete within 24 hours of the code deploy.
-2. One or more autoscaling Workers will spin up to immediately start executing new tasks based on your latest code. These Celery Workers do not wait for your previous Workers to terminate.
+1. Tasks that are `running` will continue to execute on existing Celery workers and will not be interrupted unless the task does not complete within 24 hours of the code deploy.
+2. One or more new worker(s) will spin up alongside your existing workers and immediately start executing scheduled tasks based on your latest code.
```
Comment (paolaperaza) [inline cloud/deploy-code.md:66]:
```suggestion
2. One or more new worker(s) will spin up alongside your existing workers and immediately start executing scheduled tasks based on your latest code. These Celery workers do not wait for your previous workers to terminate.

This is true for downstream tasks of DAG runs that are in-progress. For example, if you deploy to Astronomer when `Task A` of your DAG is running, `Task A` will continue to run on an old Celery worker. If `Task B` and `Task C` of that same DAG need to be run after `Task A`, they will both be scheduled on new Celery workers. To protect against DAG run failures that are due to a change in source code, we recommend implementing at least two retries at the DAG level as described in the [DAG Writing Best Practices Airflow Guide](https://www.astronomer.io/guides/dag-best-practices).
```

**Review by paolaperaza — APPROVED** [review]
@jwitz I thought we could be more explicit here in talking about tasks from shared DAG run - let me know what you think. This LGTM generally.

File: cloud/deploy-code.md (hunk 1 above)
Comment (jwitz) [reply cloud/deploy-code.md:66]: I rewrote this to separate the technical explanation from the warning/ recommendation.

**Comment by aspain** [comment]
> @aspain Is the statement that DAG Runs could fail because of a change in source code correct? I think our recommendation on retries is sound, but curious to confirm if our framing here is correct

yes that is possible - if the source code of the dag changed and the task fails, my guess is it would continue failing even with retries in place (because of the logic/code change no longer lining up with what the original DAG run expected)

The other scenario is the worker termination grace period running out and failing the task

**Comment by paolaperaza** [comment]
Hm ok that actually sounds like tasks are more likely to fail because of actual change in the DAG than because of anything to do with the code deploy or the change in workers - just updated my suggestion above to fit that. Still recommending retries but framing it for a slightly different scenario.

**Comment by jwitz** [comment]
Based on the comments here, I'm going to merge my draft that assumes the following:

- DAG runs can fail because new source code is deployed and ran on downstream tasks mid-run
- To minimize these failures, you can add DAG level retries. These retries will run the entire DAG using the new workers, meaning that there will be no discrepancies in code between tasks. 

We can continue to tweak the language here as we go, but I think that the technical details here are accurate (and their implications are clear)

**Comment by aspain** [comment]
@cmarteepants This is something we spoke about a few weeks back, I was wondering if you could take a look at the comments above and confirm this is correct?

**Comment by jwitz** [comment]
@aspain Ah got it, I misunderstood `DAG-level retries` to be a retry of the entire DAG, rather than a retry setting that applies to all tasks in a DAG. In that case, is the only way to retry a failed DAG run to clear its run status as described in this guide? https://www.astronomer.io/guides/rerunning-dags#rerunning-tasks

**Comment by aspain** [comment]
@jwitz correct, dag-level retries will just apply that number of retries to each individual task rather than retrying the entire DAG. And yes they'd have to clear the run to re-run that dagrun. Or they can re-trigger/wait for the next scheduled run

**Comment by cmarteepants** [comment]
> @cmarteepants This is something we spoke about a few weeks back, I was wondering if you could take a look at the comments above and confirm this is correct?

@aspain - yes, it's fine. @paolaperaza is also right - if the DAG is running, the risks are more about code change management than using celery:

1. The code for the tasks that are yet to run is broken somehow - which is always a risk;
2. Your DAG structure changes drastically, and the tasks allocated to the new worker no longer have a valid upstream (which is always a risk - and you should probably restart the DAG);
3. The grace period ends before the task finishes (retries will help);

**Pull request merged**